### PR TITLE
Remove `with torch.cuda.device_of`, which doesn't do what we thought

### LINKED ISSUE #431: simple_seq2seq error, lstm hidden state is not moved to cuda
I tried to run simple_seq2seq experiment and everything is working fine unless I set cude_device to 0 from -1: 
    
    python -m allennlp.run train ./tests/fixtures/encoder_decoder/simple_seq2seq/experiment.json --serialization_dir /tmp/test

here is the error I receive:
`  0%|          | 0/1 [00:00<?, ?it/s]2017-10-21 18:46:13,638 - INFO - allennlp.training.trainer - Training
Traceback (most recent call last):
  File "/informatik2/wtm/home/lakomkin/anaconda3/envs/summarization/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/informatik2/wtm/home/lakomkin/anaconda3/envs/summarization/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/run.py", line 13, in <module>
    main(prog="python -m allennlp.run")
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/commands/__init__.py", line 61, in main
    args.func(args)
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/commands/train.py", line 63, in _train_model_from_args
    train_model_from_file(args.param_path, args.serialization_dir)
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/commands/train.py", line 83, in train_model_from_file
    return train_model(params, serialization_dir)
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/commands/train.py", line 168, in train_model
    trainer.train()
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/training/trainer.py", line 370, in train
    train_metrics = self._train_epoch(epoch)
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/training/trainer.py", line 221, in _train_epoch
    loss = self._batch_loss(batch, for_training=True)
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/training/trainer.py", line 176, in _batch_loss
    output_dict = self._forward(batch, for_training=for_training)
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/training/trainer.py", line 411, in _forward
    return self._model.forward(**tensor_batch)
  File "/informatik2/wtm/home/lakomkin/distr/allennlp/allennlp/models/encoder_decoders/simple_seq2seq.py", line 159, in forward
    (decoder_hidden, decoder_context))
  File "/informatik2/wtm/home/lakomkin/anaconda3/envs/summarization/lib/python3.6/site-packages/torch/nn/modules/module.py", line 224, in __call__
    result = self.forward(*input, **kwargs)
  File "/informatik2/wtm/home/lakomkin/anaconda3/envs/summarization/lib/python3.6/site-packages/torch/nn/modules/rnn.py", line 608, in forward
    self.bias_ih, self.bias_hh,
  File "/informatik2/wtm/home/lakomkin/anaconda3/envs/summarization/lib/python3.6/site-packages/torch/nn/_functions/rnn.py", line 28, in LSTMCell
    return state(igates, hgates, hidden[1]) if b_ih is None else state(igates, hgates, hidden[1], b_ih, b_hh)
  File "/informatik2/wtm/home/lakomkin/anaconda3/envs/summarization/lib/python3.6/site-packages/torch/nn/_functions/thnn/rnnFusedPointwise.py", line 78, in forward
    cx, hy, cy)
TypeError: CudaLSTMFused_updateOutput received an invalid combination of arguments - got (int, torch.cuda.FloatTensor, torch.cuda.FloatTensor, torch.cuda.FloatTensor, torch.cuda.FloatTensor, torch.FloatTensor, torch.cuda.FloatTensor, torch.cuda.FloatTensor), but expected (int state, torch.cuda.FloatTensor input, torch.cuda.FloatTensor hidden, [torch.cuda.FloatTensor bias1 or None], [torch.cuda.FloatTensor bias2 or None], torch.cuda.FloatTensor cx, torch.cuda.FloatTensor hy, torch.cuda.FloatTensor cy)`

Maybe the problem is that LSTM's hidden state is not moved to cuda, but I am not sure.

### PR DESCRIPTION
**Pull request by matt-gardner**
Fixes #431.

We had originally used `with torch.cuda.device_of(some_variable)` to create new variables, thinking that this would create them on the save device as `some_variable`.  Turns out this doesn't actually work.  This fixes the last place in the code where we still used this pattern (I think...).

### CODE REVIEW

File: allennlp/models/encoder_decoders/simple_seq2seq.py [hunk 1]
```diff
@@ -135,8 +136,7 @@ def forward(self,  # type: ignore
         else:
             num_decoding_steps = self._max_decoding_steps
         decoder_hidden = final_encoder_output
-        with torch.cuda.device_of(decoder_hidden):
-            decoder_context = Variable(torch.zeros(1, self._decoder_output_dim))
+        decoder_context = Variable(source_mask.data.new().resize_(batch_size, self._decoder_output_dim).fill_(0))
```
Comment (pdasigi) [inline allennlp/models/encoder_decoders/simple_seq2seq.py]: I see that `resize_` also truncates or expands tensors if needed.

File: allennlp/models/encoder_decoders/simple_seq2seq.py [hunk 2]
```diff
@@ -148,9 +148,7 @@ def forward(self,  # type: ignore
                 if timestep == 0:
                     # For the first timestep, when we do not have targets, we input start symbols.
                     # (batch_size,)
-                    with torch.cuda.device_of(final_encoder_output):
-                        input_choices = Variable(torch.LongTensor([self._start_index]).expand_as(
-                                final_encoder_output[:, 0]))
+                    input_choices = Variable(source_mask.data.new().resize_(batch_size).fill_(self._start_index))
```
Comment (pdasigi) [inline allennlp/models/encoder_decoders/simple_seq2seq.py]: Is there a way to write a test to see if this creates tensors on the same device?

Comment (matt-gardner) [reply allennlp/models/encoder_decoders/simple_seq2seq.py]: The existing tests show that this works on the CPU, and we know that this works on the GPU also.  We have a hard time automatically testing things on GPUs, though, which is why this bug made it through in the first place.

File: allennlp/models/encoder_decoders/simple_seq2seq.py (hunk 1 above)
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/simple_seq2seq.py]: Yes, that's the point.

File: allennlp/models/encoder_decoders/simple_seq2seq.py (hunk 2 above)
Comment (pdasigi) [reply allennlp/models/encoder_decoders/simple_seq2seq.py]: Okay. This comment seems to indicate this should work: https://github.com/pytorch/pytorch/blob/master/torch/tensor.py#L13-L15

I didn't go through the CPP code to understand what's happening though.

Comment (matt-gardner) [reply allennlp/models/encoder_decoders/simple_seq2seq.py]: This is the recommended way to do these operations: https://discuss.pytorch.org/t/is-it-possible-to-make-a-function-transparent-between-cpu-and-gpu/1235/7.  Though apparently we missed that you pass the shape directly to `new`...

**Pull request merged**